tests/components/dhcp/test_init.py

```diff
@@ -199,6 +200,52 @@ async def test_dhcp_match_hostname_and_macaddress(hass: HomeAssistant) -> None:
     )
 
 
+async def test_dhcp_match_hostname_and_macaddress_specific_interfaces(
+    hass: HomeAssistant,
+) -> None:
+    """Test matching based on hostname and macaddress."""
+    integration_matchers = dhcp.async_index_integration_matchers(
+        [{"domain": "mock-domain", "hostname": "connect", "macaddress": "B8B7F1*"}]
+    )
+
+    def _generate_mock_adapters():
+        mock_eth0 = Mock(spec=ifaddr.Adapter)
+        mock_eth0.nice_name = "eth0"
+        mock_eth0.ips = [ifaddr.IP("192.168.0.1", 24, "eth0")]
+        mock_eth0.index = 1
+        mock_eth1 = Mock(spec=ifaddr.Adapter)
+        mock_eth1.nice_name = "eth1"
+        mock_eth1.ips = [ifaddr.IP("192.168.1.1", 24, "eth1")]
+        mock_eth1.index = 2
+        return [mock_eth0, mock_eth1]
+
+    packet = Ether(RAW_DHCP_REQUEST)
+
+    with patch(
+        "homeassistant.components.network.util.ifaddr.get_adapters",
+        return_value=_generate_mock_adapters(),
+    ):
+        async_handle_dhcp_packet = await _async_get_handle_dhcp_packet(
+            hass, integration_matchers
+        )
+    with patch.object(hass.config_entries.flow, "async_init") as mock_init:
+        await async_handle_dhcp_packet(packet)
+        # Ensure no change is ignored
+        await async_handle_dhcp_packet(packet)
+
+    assert len(mock_init.mock_calls) == 1
+    assert mock_init.mock_calls[0][1][0] == "mock-domain"
+    assert mock_init.mock_calls[0][2]["context"] == {
+        "discovery_key": DiscoveryKey(domain="dhcp", key="b8b7f16db533", version=1),
+        "source": config_entries.SOURCE_DHCP,
+    }
+    assert mock_init.mock_calls[0][2]["data"] == dhcp.DhcpServiceInfo(
+        ip="192.168.210.56",
+        hostname="connect",
+        macaddress="b8b7f16db533",
+    )
```

Comment:
Here we want to patch `aiodhcpwatcher.async_start` and assert its called with the correct indicies